src/engine/GroupBy.cpp

```diff
@@ -740,81 +771,70 @@ bool GroupBy::computeGroupByForSortedSubtree(IdTable* result,
   evaluationContext._isPartOfGroupBy = true;
 
   size_t blockSize = 65536;
-
   auto numBlocks = (size_t) ceil(((double) subresult.size()) / blockSize);
 
-  size_t overallResults = 0;
+  size_t rowIndex = 0;
 
   for (size_t i = 0; i < numBlocks; ++i) {
     evaluationContext._beginIndex = i * blockSize;
     evaluationContext._endIndex = std::min((i + 1) * blockSize, subresult.size());
 
     auto currentBlockSize = evaluationContext._endIndex - evaluationContext._beginIndex;
 
-    std::vector<sparqlExpression::detail::IntOrDouble> resultValues;
+    vector<sparqlExpression::SparqlExpression*> results;
 
-    // Is this slower than emplace_back?
-    // Calling IntOrDouble constructor twice VS. resizing of vector on the go
-    resultValues.resize(currentBlockSize);
+    size_t aggregateIndex = 0;
+    auto currentRowIndex = rowIndex;
 
-    // Evaluate child expression on block
-    auto expr = aggregates.at(0)._expression.getPimpl();
-    auto exprChildren = expr->children();
+    for (auto & aggregate : aggregates) {
+      // Evaluate child expression on block
+      auto expr = aggregate._expression.getPimpl();
 
-    sparqlExpression::ExpressionResult expressionResult =
-        exprChildren[0]->evaluate(&evaluationContext);
+      // Only allow AVG for now
+      if (dynamic_cast<sparqlExpression::AvgExpression*>(expr) == nullptr)
+        AD_FAIL();
```

Comment:
This should also be checked (maybe a more sophisticated version of it) BEFORE you actually enable the optimization.
And another check that should be done, is that none of the children of the average is itself an aggregate.